ocaml/parsing/jane_syntax.ml

```diff
@@ -284,29 +302,305 @@ module Immutable_arrays = struct
+            "Expected the alloc mode to be indicated on a Pexp_coerce or \
+             Pexp_constraint node within an n-ary function: one of \
+             %a or %a."
+            Extension_node.format (Extension_node.Alloc_mode Global)
+            Extension_node.format (Extension_node.Alloc_mode Local)
+      | Misannotated_function_cases ->
+          Location.errorf ~loc
+            "%a may be applied only to a function expression."
+            Extension_node.format Extension_node.End_cases
+      | Bad_syntactic_arity_embedding suffix ->
+          Location.errorf ~loc
+            "Unknown syntactic-arity extension point %a."
+            Embedded_name.pp_quoted_name
+            (Embedded_name.of_feature feature suffix)
+
+    exception Error of Location.t * error
+
+    let () =
+      Location.register_error_of_exn
+        (function
+          | Error(loc, err) -> Some (report_error ~loc err)
+          | _ -> None)
+
+    let raise expr err = raise (Error (expr.pexp_loc, err))
+  end
+
+  let expand_n_ary_expr expr =
+    match find_and_remove_jane_syntax_attribute expr.pexp_attributes with
+    | None -> None
+    | Some (ext_name, attributes) -> begin
+        let expr = { expr with pexp_attributes = attributes } in
+        match Jane_syntax_parsing.Embedded_name.components ext_name with
+        | feature :: suffix
+          when String.equal feature extension_string -> begin
+            match Extension_node.of_extension_suffix suffix with
+            | Some ext -> Some (ext, expr)
+            | None ->
+              Desugaring_error.raise
+                expr
+                (Bad_syntactic_arity_embedding suffix)
+          end
+        | _ :: _ ->
+          Desugaring_error.raise
+            expr
+            (Non_syntactic_arity_embedding ext_name)
+      end
+
+  let expand_n_ary_expr_with_payload expr
+      : (Extension_node_with_payload.t * Parsetree.expression) option
+    =
+    match expand_n_ary_expr expr with
+    | None -> None
+    | Some (Param, expr) -> Some (Param, expr)
+    | Some (End_expression_body, expr) -> Some (End_expression_body, expr)
+    | Some (End_cases, ({ pexp_desc = Pexp_function cases } as expr)) ->
+        Some (End_cases cases, expr)
+    | Some (End_cases, expr) ->
+        Desugaring_error.raise expr Misannotated_function_cases
+    | Some (Alloc_mode alloc_mode, expr) -> Some (Alloc_mode alloc_mode, expr)
+
+  let check_end expr ~rev_params ~function_constraint =
+    match expand_n_ary_expr_with_payload expr with
+    | Some (Param, _) | Some (Alloc_mode _, _) -> None
+    | Some (End_cases cases, expr) ->
+        let { pexp_loc = loc; pexp_attributes = attrs } = expr in
+        let params = List.rev rev_params in
+        Some (params, function_constraint, Pfunction_cases (cases, loc, attrs))
+    | Some (End_expression_body, expr) ->
+        let params = List.rev rev_params in
+        Some (params, function_constraint, Pfunction_body expr)
+    | None ->
```

Comment:
We cannot, as this pattern shadows `expr`.

Reply:
(Note: in the future this could be done with the pattern matchings changes + allowing `when` under or-patterns: `Some (End_expression_body, expr) | (None when expr match expr) -> ...`)